keyserver/server_test.go

```diff
@@ -237,3 +238,249 @@ func TestRegisterDuplicateUserInDifferentEpoches(t *testing.T) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		if str.Epoch != 0 {
+			t.Fatal("Expect STR with epoch", 0)
+		}
+		if response.AP == nil {
+			t.Fatal("Expect a proof of absence")
+		}
+		if response.TB == nil {
+			t.Fatal("Expect a TB")
+		}
+	})
+}
+
+func TestRegisterAndLookup(t *testing.T) {
+	util.WithDB(func(db kv.DB) {
+		server, teardown := startServer(t, db, 1, "")
+		defer teardown()
+
+		_, err := testutil.NewClient(t, testutil.LocalConnection, []byte(registrationMsg))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		server.dir.Update()
+		rev, err := testutil.NewClient(t, testutil.PublicConnection, []byte(keylookupMsg))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		type expectingResponse struct {
+			Type  int
+			Error ErrorCode
+			AP    json.RawMessage
+			STR   json.RawMessage
+			TB    json.RawMessage
+		}
+		var response expectingResponse
+		err = json.Unmarshal(rev, &response)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if response.Type != KeyLookupType {
+			t.Fatal("Expect a key lookup response", "got", response.Type)
+		}
+		if response.Error != Success {
+			t.Fatal("Expect no error", "got", response.Error)
+		}
+		if response.STR == nil {
+			t.Fatal("Expect the latets STR")
+		}
+		type expectingSTR struct {
+			Epoch uint64
+		}
+		var str expectingSTR
+		err = json.Unmarshal(response.STR, &str)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if str.Epoch == 0 {
+			t.Fatal("Expect STR with epoch > 0")
+		}
+		if response.AP == nil {
+			t.Fatal("Expect a proof of inclusion")
+		}
+		if response.TB != nil {
+			t.Fatal("Expect returned TB is nil")
+		}
+	})
+}
+
+func TestKeyLookup(t *testing.T) {
+	util.WithDB(func(db kv.DB) {
+		server, teardown := startServer(t, db, 60, "")
+		defer teardown()
+
+		_, err := testutil.NewClient(t, testutil.LocalConnection, []byte(registrationMsg))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		server.dir.Update()
+		rev, err := testutil.NewClient(t, testutil.PublicConnection, []byte(keylookupMsg))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		type expectingResponse struct {
+			Type  int
+			Error ErrorCode
+			AP    json.RawMessage
+			STR   json.RawMessage
+			TB    json.RawMessage
+		}
+		var response expectingResponse
+		err = json.Unmarshal(rev, &response)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if response.Type != KeyLookupType {
+			t.Fatal("Expect a key lookup response", "got", response.Type)
+		}
+		if response.Error != Success {
+			t.Fatal("Expect no error", "got", response.Error)
+		}
+		if response.STR == nil {
+			t.Fatal("Expect the latets STR")
+		}
+		type expectingSTR struct {
+			Epoch uint64
+		}
+		var str expectingSTR
+		err = json.Unmarshal(response.STR, &str)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if str.Epoch == 0 {
+			t.Fatal("Expect STR with epoch > 0")
+		}
+		if response.AP == nil {
+			t.Fatal("Expect a proof of inclusion")
+		}
+		if response.TB != nil {
+			t.Fatal("Expect returned TB is nil")
+		}
+	})
+}
+
+func TestKeyLookupInEpoch(t *testing.T) {
```

Comment:
In go 1.7 the tls behaviour changed: https://golang.org/doc/go1.7#minor_library_changes (scroll to crypto/tls)
I experimented with `DynamicRecordSizingDisabled` to enforce the "old" 1.6 behaviour but the `Read` in `testutil.NewClient` still doesn't read enough data. In general it would be right to not rely on the on the underlying implementation of `net.Conn` and to make sure that `Read` and `Write` have received and written the expected amount of bytes.